src/anyp/TrafficMode.h

```diff
@@ -73,12 +73,93 @@ class TrafficMode
      */
     bool tunnelSslBumping = false;
 
-    /** true if the traffic is in any way intercepted
-     *
-     */
-    bool isIntercepted() { return natIntercept||tproxyIntercept ;}
+    PortKind portKind; ///< the parsed port type value
 };
 
+/**
+ * Set of 'mode' flags defining types of traffic which can be received.
+ *
+ * Use to determine the processing steps which need to be applied
+ * to this traffic under any special circumstances which may apply.
+ */
+class TrafficMode
+{
+public:
+    explicit TrafficMode(const TrafficModeFlags::PortKind aPortKind) : flags_(aPortKind) {}
+
+    /// This is a gateway (a.k.a. reverse proxy) port. TODO: Rename.
+    bool accelSurrogate() const { return flags_.accelSurrogate; }
+
+    /// This port handles intercepted traffic. Traffic may be intercepted many
+    /// times (at various locations) before reaching this port, including:
+    /// * between the user agent and the connecting TCP client and/or
+    /// * between the connecting TCP client and this Squid port.
+    /// This port mode alone does not imply that the client of the accepted TCP
+    /// connection was not connecting directly to this port (since commit
+    /// 151ba0d). In other words, we may not be an interception proxy ourselves.
+    bool interceptedSomewhere() const { return flags_.natIntercept || flags_.tproxyIntercept || proxySurrogateHttpsSslBump(); }
+
+    /// The user agent was explicitly configured to use a proxy at this port. We
+    /// are configured to assume that no interception has happened on the way.
+    /// This port configuration is also known as a forward proxy.
+    bool explicitProxy() const { return !interceptedSomewhere() && !accelSurrogate(); }
+
+    /// whether the PROXY protocol header is required
+    bool proxySurrogate() const { return flags_.proxySurrogate; }
+
+    /// The client of the accepted connection was not connecting to this port,
+    /// but Squid used NAT interception to accept the client connection.
+    /// The accepted traffic may have been intercepted earlier as well!
+    bool natInterceptLocally() const { return flags_.natIntercept && !proxySurrogate(); }
```

Comment:
> proxySurrogate() and NAT are unconnected states. It is possible to NAT intercept a TCP connection with PROXY on it.

Agreed but this method does not state that. It is used(or should be used) in situations when one option is on and the other is off.

Reply:
Why are such situations even being considered? they are so unrelated that no code should even be checking for that case.

Reply:
These situations are considered because they exist For example, FtpServer.cc, which does not support PROXY protocol.

Reply:
Ah. That FtpServer case is a caller-specific condition working around bugs in the current implementation of PROXY protocol. So, it should not be embedded into the TrafficMode API since that pollutes other checks that only care about the NAT part.

Reply:
> Why are such situations even being considered?

Please see my [answer](https://github.com/squid-cache/squid/pull/835/files#r825494994) to the similar request.